### DISCUSSION TITLE
[Realtime] Subscribe to the Same Channel in Different Components

### DISCUSSION THREAD

**Discussion by mrcaidev** [supabase/supabase#27142] [Questions] [answered]
Is there any live example where supabase subscribes to the same realtime channel in different components?

For example, I want to subscribe to "room-1" in `<Chat />` child component to receive chat messages, and also to "room-1" in `<Room />` parent component to sync user presence.

```tsx
<Room>     {/** Subscribe to room-1 */}
  <Chat /> {/** Also subscribe to room-1 */}
</Room>
```

I'm trying to use this feature, but the child component `<Chat />` always fails to subscribe. A log of events happened:

![image](https://github.com/supabase/supabase/assets/78269445/3f13dde3-ed8f-4fdb-87d5-448d76f23e84)

And I have also tried to share a `supabase` client with React ref, but this doesn't work either:

![image](https://github.com/supabase/supabase/assets/78269445/eb0fdb7c-d273-43b9-ad2c-c530c76efd61)

I see from the doc that it should have been pretty straightforward, but I just can't get it done. Any help would be appreciated!

**Comment by GaryAustin1** [comment]
An issue I've seen with users using realtime with React is they have useEffect's  starting realtime connections and an unsubscribe as part of that.  But they use the same channel name for them.  When paging occurs a new channel connection starts with the same name while the old one is being unsubscribed.   The two realtime websocket calls overlap and the new connection gets messed up with the old connection unsubscribing.   The solution is to move the realtime to a higher level or not use the same channel names.

You don't really show any code, so just a wild guess.

**Reply by mrcaidev** [reply]
Thanks! Yes, I do subscribe & unsubscribe in `useEffect`, and I guess the overlap you mentioned is probably the root of the problem. But I still have some questions:

One is that, even if I move the channel subscription to a higher level, don't I still need to resort to `useEffect` for subscription/unsubscription? And [the official demo](https://github.com/supabase/realtime/blob/main/demo/pages/%5B...slug%5D.tsx#L479-L480) also uses this approach. Why is this not causing any overlap?

Another is that, when I move the channel higher, do I need to register all events in that higher component, and also all relevant states and handlers as a result, just like the official demo do? I'm trying to break it down to smaller pieces for maintenance.

Here is my code in `<Chat />`:

```tsx
useEffect(() => {
    const supabase = createBrowserSupabase();

    const channel = supabase.channel(channelName, {
      config: { broadcast: { ack: true, self: true } },
    });

    channel
      .on<ChatMessage>("broadcast", { event: "chat_message" }, (payload) => {
        setChatEntries((prev) => [...prev, payload.payload]);
      })
      .on<User>("presence", { event: "join" }, ({ newPresences }) => {
        const newUser = newPresences[0]!;
        setChatEntries((prev) => [...prev, { type: "JOIN", user: newUser }]);
      })
      .on<User>("presence", { event: "leave" }, ({ leftPresences }) => {
        const leftUser = leftPresences[0]!;
        setChatEntries((prev) => [...prev, { type: "LEAVE", user: leftUser }]);
      })
      .subscribe(async (status, error) => {
        if (status !== "SUBSCRIBED") {
          return;
        }

        channelRef.current = channel; // For functions outside the effect to use the channel.
      });

    return () => {
      supabase.removeChannel(channel);
    };
  }, [channelName]);
```

**Reply by GaryAustin1** [reply, marked as answer]
I unfortunately don't use React myself, so only know of the issue from helping others that do.  I can't guide you on the proper solution other than don't share the same channel name in multiple places if it can be unsubscribed.  You can have many channels at no "cost" as they all share the same connection.

**Reply by mrcaidev** [reply]
Thank you! I have some rough ideas about how to refactor now.

**Reply by joshuabaker** [reply]
@GaryAustin1 Just to weigh in here, this needs a fuller response from Supabase, or a specific section in the documentation. 

The stateful nature of React makes subscribing to something more than a single stream on a single channel name extremely complicated (perhaps impossible). There’s no error thrown here either, so it’s days of work lost trying to fight around how the client disallows a common event listener pattern.

@mrcaidev Did you just end up having different streams for each table/component/case?

**Reply by GaryAustin1** [reply]
@joshuabaker Supabase is not likely to read this discussion.  If you have feedback for them (I'm not an employee, but a user helping out) then you should generate an issue in probably the supabase/realtime respository.
